Add warning if playbook references a var that shadows a builtin method

alikins opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/template/__init__.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (warn_on_invalid_var_getattr 3ecac85b92) last updated 2017/03/03 16:27:40 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = Default w/o overrides

```

##### SUMMARY
There is a silent gotcha lurking in http://docs.ansible.com/ansible/playbooks_variables.html#what-makes-a-valid-variable-name

If playbook yaml data structures contain fields with any ~80 var names listed there, and the playbook references them via dot/getattr access ( "{{ my_data.my_field }}"), instead of returning the data item set in the data structure, a reference to the python builtin methods of the same names will be returned.

There is a way to avoid that (using {{ my_data['my_field'] }} to prefer the getitem()), but it is very easy to accidentally use a shadowed builtin and the getattr/dot access method and end up with a the method reference which caused very confusing error messages and other weird failures.
ie, if my_data:

``` yaml
my_data:
   get: http://www.ansible.com
   update: http://www.ansible.com/releases
   blippy: 37
```

If playbook uses '{{ my_data.get }}', the result   will be something like '<built-in method get of dict object at 0x7fdef81cc050>' instead of 'http://www.ansible.com'

This pr adds a warning message for that case. It also changes the behavior to prefer the data item ('http://www.ansible.com') for cases where data items shadow built in methods. A reference to
'{{ my_data.popitem }}' will still return the method reference (popitem is not shadowed).

The patch detects if a getattr is used on the object could also reference a data item. If so, it warns and returns the data item.

alikins committed:
warn if a template var getattr (var.some_attr) returns a method

alikins committed:
remove unused

alikins committed:
if we getattr a method, use getitem instead.

Warn on dangerous use and DWIM.

alikins committed:
cleanup

alikins committed:
cleanup

alikins committed:
add asserts to intg tests

mattclay commented:
CI failure in integration tests due to:

```
2017-03-03 21:45:45 TASK [template_getattr : Try to access and call valid_list.confused_for_list.popitem (builtin method) and expect the method to be accessed and called without a warning] ***
2017-03-03 21:45:45 task path: /root/ansible/test/integration/targets/template_getattr/tasks/main.yml:164
2017-03-03 21:45:45 fatal: [testhost]: FAILED! => {
2017-03-03 21:45:45     "assertion": "valid_test.confused_for_a_list.popitem()[0] == \"pop\"",
2017-03-03 21:45:45     "changed": false,
2017-03-03 21:45:45     "evaluated_to": false,
2017-03-03 21:45:45     "failed": true
2017-03-03 21:45:45 }
```

Also, the test `ansible-test sanity --test yamllint` failed with the following errors:

```      
test/integration/targets/template_getattr/tasks/main.yml:145:9: too many spaces after hyphen (hyphens)
test/integration/targets/template_getattr/tasks/main.yml:179:1: too many blank lines (1 > 0) (empty-lines)
```

alikins committed:
Fix intg test failure (assumed dict order for popitem())

alikins committed:
yamllint cleanup

alikins committed:
pep8 cleanup

abadger commented:
The warning seems like a good idea as the original error is a bit cryptic but changing the behaviour has the potential to break things.  here's a bit of a contrived example:

```
---
- hosts: localhost
  gather_facts: False
  vars:
    test:
      update: data
    test2:
      other: 1
  tasks:
  - debug:
      msg: '{{ test.update(test2) }}{{test}}'
```

Ah, here's a more useful example:
```
---
- hosts: localhost
  gather_facts: False
  vars:
    test:
      keys: data
    test2:
      other: 1
  tasks:
  - debug:
      msg: '{{ test.keys() }}'
```

alikins commented:
reference a objects method attrs result in getting the repr of the method in devel/. 
Even if there is also a data item with the same name (ie, 'update')

```
[newswoop:F25:ansible (devel % u=)]$ cat test_whatever2.yml 
---
- hosts: localhost
  gather_facts: False
  vars:
    test:
      update: data
    test2:
      other: 1
  tasks:
  - debug:
      msg: "test.update: {{ test.update }}"
  - debug:
      msg: "test2.update: {{test2.update}}"
[newswoop:F25:ansible (devel % u=)]$ ansible-playbook -vvv test_whatever2.yml 
ansible-playbook 2.4.0 (devel 25a9ababcc) last updated 2017/08/17 15:20:13 (GMT -400)
  config file = /home/adrian/src/ansible/ansible.cfg
  configured module search path = [u'/home/adrian/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/adrian/src/ansible/lib/ansible
  executable location = /home/adrian/src/ansible/bin/ansible-playbook
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
Using /home/adrian/src/ansible/ansible.cfg as config file
 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: Could not match supplied host pattern, ignoring: all

 [WARNING]: provided hosts list is empty, only localhost is available


PLAYBOOK: test_whatever2.yml *********************************************************************
1 plays in test_whatever2.yml

PLAY [localhost] *********************************************************************************
META: ran handlers

TASK [debug] *************************************************************************************
task path: /home/adrian/src/ansible/test_whatever2.yml:10
ok: [localhost] => {
    "msg": "test.update: <built-in method update of dict object at 0x7f5df6250050>"
}

TASK [debug] *************************************************************************************
task path: /home/adrian/src/ansible/test_whatever2.yml:12
ok: [localhost] => {
    "msg": "test2.update: <built-in method update of dict object at 0x7f5df6255280>"
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0   
```

alikins committed:
Warn on template ref a shadowed method but return it anyway.

Display a warning if a template references a attribute name
that is a object method and also an item in the data.

In this ambiquous scenario, dislay the warning, then return the
object method and not the data item.

abadger commented:
> reference a objects method attrs result in getting the repr of the method in devel/.
Even if there is also a data item with the same name (ie, 'update')

No, it results in getting the **method**.  you see the repr when you attempt to print the method.  but most of the time you want to use the method, not print it.  We can't tell in the getattr how the method is going to be used.  My example of '{{ mydict.keys() }}' is a perfectly valid thing to want to do and with your patch that would be broken.

bcoca also raises the question of performance.  We template quite a lot and the templating is already slow.  Intercepting getattr would slow that further.  We'd want to benchmark the overhead of adding the warning when there's many variables before we push just the warning code in.

alikins committed:
A couple of test cases for string methods

mattclay commented:
bot_status
